File: benchmarks/hawkes_simulation.py
import time
from benchmarks.create_hawkes_data import get_parameters

from tick.simulation import SimuHawkesExpKernels, SimuHawkesMulti


if __name__ == "__main__":
    n_simulations = 100

    n_nodes_sample = [1, 2, 4, 16]
    end_times = [1000, 2000, 5000, 10000, 20000, 50000, 100000]

    for n_nodes in n_nodes_sample:
        decays, baseline, adjacency = get_parameters(n_nodes)

        if False:
            print("\nn nodes", n_nodes)
            print('baseline\n', baseline)
            print('decays\n', decays)
            print('adjacency\n', adjacency)
            print('adjacency * decays\n', adjacency * decays)

        for end_time in end_times:
            hawkes_exp_kernels = SimuHawkesExpKernels(
                adjacency=adjacency, decays=decays, baseline=baseline,
                end_time=end_time, seed=1039, verbose=False)

            start_time = time.time()
            n_events = 0
            for _ in range(n_simulations):
                hawkes_exp_kernels.reset()
                hawkes_exp_kernels.simulate()
                n_events += hawkes_exp_kernels.n_total_jumps
            simulation_time = time.time() - start_time

            if False:
                print("Simulating {} events costs {} secs"
                      .format(n_events, simulation_time))
            else:
                print("simulation,tick 1,{},{},{:.6f}"
                      .format(n_nodes, n_events, simulation_time))

File: benchmarks/hawkes_simulation.R
# might need to run
# install.packages("hawkes")

library(hawkes)

source("hawkes_data.R")


simulate <- function (lambda0, alpha, beta, end_time, n_simulations) {
    n_total_jumps <- 0;
    for (i in 1:n_simulations) {
        history <- simulateHawkes(lambda0, alpha, beta, end_time)
        n_total_jumps <- n_total_jumps + Reduce("+", lapply(history, length))
    }
    return(n_total_jumps)
}

n_simulations = 100
n_nodes_sample <- c(1, 2, 4, 16)
end_times <- c(1000, 2000, 5000, 10000, 20000, 50000, 100000)


for (n_nodes in n_nodes_sample) {
    for (end_time in end_times) {
        parameters <- get_parameters(n_nodes)
        beta <- parameters[[1]]
        lambda0 <- parameters[[2]]
        alpha <- parameters[[3]]

        benchmark <- system.time(
            n_total_jumps <- simulate(lambda0, alpha, beta, end_time, n_simulations))

        simulation_time <- benchmark[["elapsed"]]
        # print(sprintf("Simulating %i events costs %f secs",
        #               n_total_jumps, simulation_time))
        cat(sprintf('simulation,hawkes R,%i,%i,%.5f\n',
                    n_nodes, n_total_jumps, simulation_time))
    }
}

